src/ast/attachpoint_parser.cpp

```diff
@@ -0,0 +1,501 @@
+#include "attachpoint_parser.h"
+
+#include <algorithm>
+#include <exception>
+#include <functional>
+#include <string>
+#include <unordered_map>
+
+#include "types.h"
+
+namespace bpftrace {
+namespace ast {
+
+AttachPointParser::AttachPointParser(Program *root,
+                                     BPFtrace &bpftrace,
+                                     std::ostream &sink)
+    : root_(root), bpftrace_(bpftrace), sink_(sink)
+{
+}
+
+int AttachPointParser::parse()
+{
+  if (!root_)
+    return 1;
+
+  uint32_t failed = 0;
+  for (Probe *probe : *(root_->probes))
+  {
+    for (AttachPoint *ap : *(probe->attach_points))
+    {
+      if (parse_attachpoint(*ap))
+      {
+        ++failed;
+        bpftrace_.error(sink_, ap->loc, errs_.str());
+        errs_.str({}); // clear buffer
+      }
+    }
+  }
+
+  return failed;
+}
+
+int AttachPointParser::parse_attachpoint(AttachPoint &ap)
+{
+  ap_ = &ap;
+
+  parts_ = split_string(ap_->raw_input, ':', true);
+  if (parts_.empty())
+  {
+    errs_ << "Invalid attachpoint definition" << std::endl;
+    return 1;
+  }
+
+  ap_->provider = probetypeName(parts_.front());
+
+  switch (probetype(parts_.front()))
+  {
+    case ProbeType::kprobe:
+      return kprobe_parser();
+    case ProbeType::kretprobe:
+      return kretprobe_parser();
+    case ProbeType::uprobe:
+      return uprobe_parser();
+    case ProbeType::uretprobe:
+      return uretprobe_parser();
+    case ProbeType::usdt:
+      return usdt_parser();
+    case ProbeType::tracepoint:
+      return tracepoint_parser();
+    case ProbeType::profile:
+      return profile_parser();
+    case ProbeType::interval:
+      return interval_parser();
+    case ProbeType::software:
+      return software_parser();
+    case ProbeType::hardware:
+      return hardware_parser();
+    case ProbeType::watchpoint:
+      return watchpoint_parser();
+    default:
+      errs_ << "Unrecognized provider: " << ap_->provider << std::endl;
```

Comment:
```suggestion
      errs_ << "Unrecognized probe type: " << ap_->provider << std::endl;
```
and in other error messages